src/erlcloud_s3.erl

```diff
@@ -1084,17 +1088,95 @@ make_link(Expire_time, BucketName, Key, Config) ->
 
 -spec get_object_url(string(), string()) -> string().
 
- get_object_url(BucketName, Key) ->
+get_object_url(BucketName, Key) ->
   get_object_url(BucketName, Key, default_config()).
 
 -spec get_object_url(string(), string(), aws_config()) -> string().
 
- get_object_url(BucketName, Key, Config) ->
+get_object_url(BucketName, Key, Config) ->
   case Config#aws_config.s3_bucket_after_host of
       false -> lists:flatten([Config#aws_config.s3_scheme, BucketName, ".", Config#aws_config.s3_host, port_spec(Config), "/", Key]);
       true  -> lists:flatten([Config#aws_config.s3_scheme, Config#aws_config.s3_host, port_spec(Config), "/", BucketName, "/", Key])
   end.
 
+-spec get_object_url_elements(string(), string(), aws_config()) -> {Host::string(), Path::string(), URL::string()}.
+get_object_url_elements(BucketName, Key, Config) ->
+    Key0 = case lists:prefix("/", Key) of
+               true -> Key;
+               false -> "/" ++ Key
+           end,
+
+    case Config#aws_config.s3_bucket_after_host of
+        false ->
+            Host = BucketName ++ "." ++ Config#aws_config.s3_host,
+            {Host, Key0, lists:flatten([Config#aws_config.s3_scheme, Host, port_spec(Config), Key0])};
+        true  ->
+            Host = Config#aws_config.s3_host,
+            Path = lists:flatten(["/", BucketName, Key0]),
+            {Host, Path, lists:flatten([Config#aws_config.s3_scheme, Host, port_spec(Config), Path])}
+    end.
+
+-spec signature(aws_config(), string(), string(), string(), atom(), proplist(), proplist(), string()) -> string().
+signature(Config, Path, Date, Region, Method, QueryParams, Headers, Payload) ->
+  Service = "s3",
+  CredentialScope = erlcloud_aws:credential_scope(Date, Region, Service),
+  {CanonicalRequest, SignedHeaders} = erlcloud_aws:canonical_request(Method, Path, QueryParams, Headers, Payload),
+  ToSign = erlcloud_aws:to_sign(Date, CredentialScope, CanonicalRequest),
+  SigningKey = erlcloud_aws:signing_key(Config, Date, Region, Service),
+  [Result] = erlcloud_aws:base16(erlcloud_util:sha256_mac(SigningKey, ToSign)),
+  Result.
```

Comment:
I don't see many tests for this function. It might be nice to generate some pre-signed URLs with another tool and then use the results of that as test data that we could use to validate this function.

Reply:
I don't think this is my code.  Would you still like me to add tests for it?

Reply:
yeah - that would be really nice if we can add tests, that might also make it easier to get the commit merged upstream.

Reply:
"...generate some pre-signed URLs with another tool..."

1) Does anyone know of a trusted tool that does this?

2) In theory we could simply take presigned URLs that worked against S3 in the past (meaning, we know they are valid) and run them through again, but the problem is, they expire.  The date looks mockable so that might not be an issue.  I'd have to create a build that spits out the inputs+outputs that came into/out-of this function which were then subsequently used to hit S3 with presigned URLs successfully.

Reply:
Ok, I did #2 above.  I created a build to log diagnostics, and ran `chef-server-ctl test` against S3.  I was able to get the inputs and outputs to the signature function to create tests.  The tests worked.  However, my S3 `access_key_id` is in the source code.  Perhaps a better approach is to run against Bookshelf instead, although that is using our own code to verify our own code.

Reply:
Rewrote to use Bookshelf vs S3.